mlir/lib/Dialect/MIOpen/Transforms/ConvToGemmLowering.cpp

```diff
@@ -224,13 +244,30 @@ LogicalResult elementwiseConversion(Conv2DBwdWeightOp op, PatternRewriter &b) {
       collapsedWorkspace.getType().cast<MemRefType>().getShape();
   assert((collapsedFilterShape[0] == collapsedWorkspaceShape[0]) &&
          "Filter tensor and workspace size mismatch");
-  auto loop = b.create<AffineForOp>(loc, 0, collapsedWorkspaceShape[0]);
-  auto iv = loop.getInductionVar();
+
+  // Pseudo code:
+  // size_t offset = (hipBlockIdx_x * hipBlockDim_x + hipThreadIdx_x;
+  // size_t stride = hipBlockDim_x * hipGridDim_x;
+  // for (size_t i = offset; i < sizeof(collapsedOutput); i+= stride)
+  //     collapsedFilter[i] = convert(collapsedWorkspace[i]);
+
+  auto workgroupId = b.create<WorkgroupIdOp>(loc, b.getIndexType());
+  auto workgroupDim = b.create<ConstantIndexOp>(loc, kUtilityKernelBlockSize);
+  auto workitemId = b.create<WorkitemIdOp>(loc, b.getIndexType());
+  auto offset = b.create<AddIOp>(
+      loc, b.create<MulIOp>(loc, workgroupId, workgroupDim), workitemId);
+  auto gridDim = b.create<ConstantIndexOp>(loc, kUtilityKernelGridSize);
+  auto stride = b.create<MulIOp>(loc, workgroupDim, gridDim);
+
+  auto loop = b.create<scf::ForOp>(
+      loc, offset, b.create<ConstantIndexOp>(loc, collapsedWorkspaceShape[0]),
+      stride);
   b.setInsertionPointToStart(loop.getBody());
-  auto loadedValue = b.create<AffineLoadOp>(loc, collapsedWorkspace, iv);
+  auto iv = loop.getInductionVar();
+  auto loadedValue = b.create<memref::LoadOp>(loc, collapsedWorkspace, iv);
   auto convertedValue =
       createTypeConversionOp(b, loc, loadedValue, filterDataType);
```

Comment:
I think `elementwiseConversion` is much more generic than the `zeroInit`. I'd prefer both to use a more generic implementation such that the way we generate `convertedValue` is an argument where:
 - In `zeroInit`, the functor just set 0
 - In `elementwiseConversion`, the functor does type conversion

This also makes future elementwise kernels to be more straightforward to implement.

Reply:
Pushed a new commit to abstract the elementwise emission loop.